### ISSUE TITLE
spdk_nvme_ns_cmd_write data loss

### ISSUE THREAD

**Issue by WenhaoYuan604** [spdk/spdk#3563]
# Sighting report

I called the spdk_nvme_ns_cmd_write function to write the data from the buffer sequence.buffer to one block. 

The data in the buffer is as follows, which contains two "hello world!" strings.

snprintf(sequence.buf, 0x1000, "%s", "Hello world!\n");
char * buffer = sequence.buf + 1024;
snprintf(buffer, 0x0c00, "%s", "Hello world!\n");

Then, I called spdk_nvme_ns_cmd_read to read the block.

## Expected Behavior
The buffer of read should contain two "hello world!" strings.

## Current Behavior
The buffer of read only contain first ”hello world!“ string

## Steps to Reproduce

1.
2.
3.
4.

## Context (Environment including OS version, SPDK version, etc.)

OS:Linux 5.4.0-196-generic

SPDK：v23.05

**Comment by jimharris** [comment]
It would be helpful to show all of your code.

**Comment by WenhaoYuan604** [comment]
Here is my code.

The main modifications were made to the read complete and write complete functions.


#include "./spdk_struct.h"

int add_trid(const char *trid_str)
{
	struct trid_entry *trid_entry;
	struct spdk_nvme_transport_id *trid;
	char *ns;
	char *hostnqn;

	trid_entry = calloc(1, sizeof(*trid_entry));
	if (trid_entry == NULL) {
        printf("trid_entry allocted memory failed\n");
		return -1;
	}
	trid = &trid_entry->trid;
	trid->trtype = SPDK_NVME_TRANSPORT_PCIE;
	snprintf(trid->subnqn, sizeof(trid->subnqn), "%s", SPDK_NVMF_DISCOVERY_NQN);
	if (spdk_nvme_transport_id_parse(trid, trid_str) != 0) {
		printf("Invalid transport ID format '%s'\n", trid_str);
		free(trid_entry);
		return 1;
	}
	ns = strcasestr(trid_str, "ns:");
	if (ns) {
		char nsid_str[6]; /* 5 digits maximum in an nsid */
		int len;
		int nsid;
		ns += 3;
		len = strcspn(ns, " \t\n");
		if (len > 5) {
			printf("NVMe namespace IDs must be 5 digits or less\n");
			free(trid_entry);
			return 1;
		}
		memcpy(nsid_str, ns, len);
		nsid_str[len] = '\0';
		nsid = spdk_strtol(nsid_str, 10);
		if (nsid <= 0 || nsid > 65535) {
			fprintf(stderr, "NVMe namespace IDs must be less than 65536 and greater than 0\n");
			free(trid_entry);
			return 1;
		}
		trid_entry->nsid = (uint16_t)nsid;
	}
	hostnqn = strcasestr(trid_str, "hostnqn:");
	if (hostnqn) {
		size_t len;
		hostnqn += strlen("hostnqn:");
		len = strcspn(hostnqn, " \t\n");
		if (len > (sizeof(trid_entry->hostnqn) - 1)) {
			fprintf(stderr, "Host NQN is too long\n");
			free(trid_entry);
			return 1;
		}
		memcpy(trid_entry->hostnqn, hostnqn, len);
		trid_entry->hostnqn[len] = '\0';
	}
    // printf("insert g_trid_list %d ,trid_entry is %d\n",&g_trid_list,trid_entry);
	TAILQ_INSERT_TAIL(&g_trid_list, trid_entry, tailq);
    // printf("after insert g_trid_list %d ,g_trid_list.first is %d\n",&g_trid_list.tqh_first);
	return 0;
}

bool
probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
	 struct spdk_nvme_ctrlr_opts *opts)
{
	printf("Attaching to %s\n", trid->traddr);

	return true;
}

void
register_ns(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_ns *ns)
{
	struct ns_entry *entry;

	if (!spdk_nvme_ns_is_active(ns)) {
		return;
	}

	entry = malloc(sizeof(struct ns_entry));
	if (entry == NULL) {
		perror("ns_entry malloc");
		exit(1);
	}

	entry->ctrlr = ctrlr;
	entry->ns = ns;
	TAILQ_INSERT_TAIL(&g_namespaces, entry, link);

	printf("  Namespace ID: %d size: %juGB\n", spdk_nvme_ns_get_id(ns),
	       spdk_nvme_ns_get_size(ns) / 1000000000);
}

struct hello_world_sequence {
	struct ns_entry	*ns_entry;
	char		*buf;
	unsigned        using_cmb_io;
	int		is_completed;
};

static void
read_complete(void *arg, const struct spdk_nvme_cpl *completion)
{
	struct hello_world_sequence *sequence = arg;

	/* Assume the I/O was successful */
	sequence->is_completed = 1;
	/* See if an error occurred. If so, display information
	 * about it, and set completion value so that I/O
	 * caller is aware that an error occurred.
	 */
	if (spdk_nvme_cpl_is_error(completion)) {
		spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
		fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
		fprintf(stderr, "Read I/O failed, aborting run\n");
		sequence->is_completed = 2;
		exit(1);
	}

	/*
	 * The read I/O has completed.  Print the contents of the
	 *  buffer, free the buffer, then mark the sequence as
	 *  completed.  This will trigger the hello_world() function
	 *  to exit its polling loop.
	 */
	char * buffer = sequence->buf + 1024;
	printf("buffer is %s%s\n" , sequence->buf , buffer);

	int length = 0;
	for(int i = 0 ; i< 0x1000 ; i++){
		if(sequence->buf[i] != 0)
			length++;
	}
	printf("length is %d\n" , length);
	spdk_free(sequence->buf);
}

static void
write_complete(void *arg, const struct spdk_nvme_cpl *completion)
{
	struct hello_world_sequence	*sequence = arg;
	struct ns_entry			*ns_entry = sequence->ns_entry;
	int				rc;

	/* See if an error occurred. If so, display information
	 * about it, and set completion value so that I/O
	 * caller is aware that an error occurred.
	 */
	if (spdk_nvme_cpl_is_error(completion)) {
		spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
		fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
		fprintf(stderr, "Write I/O failed, aborting run\n");
		sequence->is_completed = 2;
		exit(1);
	}
	/*
	 * The write I/O has completed.  Free the buffer associated with
	 *  the write I/O and allocate a new zeroed buffer for reading
	 *  the data back from the NVMe namespace.
	 */
	if (sequence->using_cmb_io) {
		spdk_nvme_ctrlr_unmap_cmb(ns_entry->ctrlr);
	} else {
		spdk_free(sequence->buf);
	}
	sequence->buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);

	rc = spdk_nvme_ns_cmd_read(ns_entry->ns, ns_entry->qpair[0], sequence->buf,
				   0, /* LBA start */
				   1, /* number of LBAs */
				   read_complete, (void *)sequence, 0);
	if (rc != 0) {
		fprintf(stderr, "starting read I/O failed\n");
		exit(1);
	}
}


static void
hello_world(int num_qpair)
{
	struct ns_entry			*ns_entry;
	struct hello_world_sequence	sequence;
	int				rc;
	size_t				sz;

	TAILQ_FOREACH(ns_entry, &g_namespaces, link) {
		
		ns_entry->qpair = (struct spdk_nvme_qpair **)malloc(sizeof(struct spdk_nvme_qpair *) * num_qpair);
		for(int i = 0 ; i < num_qpair ; i++ ){
			ns_entry->qpair[i] = spdk_nvme_ctrlr_alloc_io_qpair(ns_entry->ctrlr, NULL, 0);
			if (ns_entry->qpair[i] == NULL) {
				printf("ERROR: spdk_nvme_ctrlr_alloc_io_qpair() failed\n");
				return ;
			}
		}

		sequence.using_cmb_io = 1;
		sequence.buf = spdk_nvme_ctrlr_map_cmb(ns_entry->ctrlr, &sz);
		if (sequence.buf == NULL || sz < 0x1000) {
			sequence.using_cmb_io = 0;
			sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
		}
		if (sequence.buf == NULL) {
			printf("ERROR: write buffer allocation failed\n");
			return;
		}
		if (sequence.using_cmb_io) {
			printf("INFO: using controller memory buffer for IO\n");
		} else {
			printf("INFO: using host memory buffer for IO\n");
		}
		sequence.is_completed = 0;
		sequence.ns_entry = ns_entry;

		/*
		 * If the namespace is a Zoned Namespace, rather than a regular
		 * NVM namespace, we need to reset the first zone, before we
		 * write to it. This not needed for regular NVM namespaces.
		 */
		/*
		 * Print "Hello world!" to sequence.buf.  We will write this data to LBA
		 *  0 on the namespace, and then later read it back into a separate buffer
		 *  to demonstrate the full I/O path.
		 */
		snprintf(sequence.buf, 0x1000, "%s", "Hello world!\n");
		char * buffer = sequence.buf + 1024;
		snprintf(buffer, 0x0c00, "%s", "Hello world!\n");
		printf("buffer is %s%s\n" , sequence.buf , buffer);
		int length = 0;
		for(int i = 0 ; i< 0x1000 ; i++){
			if(sequence.buf[i] != 0)
				length++;
		}
		printf("length is %d\n" , length);
		/*
		 * Write the data buffer to LBA 0 of this namespace.  "write_complete" and
		 *  "&sequence" are specified as the completion callback function and
		 *  argument respectively.  write_complete() will be called with the
		 *  value of &sequence as a parameter when the write I/O is completed.
		 *  This allows users to potentially specify different completion
		 *  callback routines for each I/O, as well as pass a unique handle
		 *  as an argument so the application knows which I/O has completed.
		 *
		 * Note that the SPDK NVMe driver will only check for completions
		 *  when the application calls spdk_nvme_qpair_process_completions().
		 *  It is the responsibility of the application to trigger the polling
		 *  process.
		 */
		rc = spdk_nvme_ns_cmd_write(ns_entry->ns, ns_entry->qpair[0], sequence.buf,
					    0, /* LBA start */
					    1, /* number of LBAs */
					    write_complete, &sequence, SPDK_NVME_IO_FLAGS_FORCE_UNIT_ACCESS);
		if (rc != 0) {
			fprintf(stderr, "starting write I/O failed\n");
			exit(1);
		}

		/*
		 * Poll for completions.  0 here means process all available completions.
		 *  In certain usage models, the caller may specify a positive integer
		 *  instead of 0 to signify the maximum number of completions it should
		 *  process.  This function will never block - if there are no
		 *  completions pending on the specified qpair, it will return immediately.
		 *
		 * When the write I/O completes, write_complete() will submit a new I/O
		 *  to read LBA 0 into a separate buffer, specifying read_complete() as its
		 *  completion routine.  When the read I/O completes, read_complete() will
		 *  print the buffer contents and set sequence.is_completed = 1.  That will
		 *  break this loop and then exit the program.
		 */
		while (!sequence.is_completed) {
			spdk_nvme_qpair_process_completions(ns_entry->qpair[0], 0);
		}

		/*
		 * Free the I/O qpair.  This typically is done when an application exits.
		 *  But SPDK does support freeing and then reallocating qpairs during
		 *  operation.  It is the responsibility of the caller to ensure all
		 *  pending I/O are completed before trying to free the qpair.
		 */
		// spdk_nvme_ctrlr_free_io_qpair(ns_entry->qpair);
	}
}


void
attach_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
	  struct spdk_nvme_ctrlr *ctrlr, const struct spdk_nvme_ctrlr_opts *opts)
{
	int nsid;
	struct ctrlr_entry *entry;
	struct spdk_nvme_ns *ns;
	const struct spdk_nvme_ctrlr_data *cdata;

	entry = malloc(sizeof(struct ctrlr_entry));
	if (entry == NULL) {
		perror("ctrlr_entry malloc");
		exit(1);
	}

	printf("Attached to %s\n", trid->traddr);

	/*
	 * spdk_nvme_ctrlr is the logical abstraction in SPDK for an NVMe
	 *  controller.  During initialization, the IDENTIFY data for the
	 *  controller is read using an NVMe admin command, and that data
	 *  can be retrieved using spdk_nvme_ctrlr_get_data() to get
	 *  detailed information on the controller.  Refer to the NVMe
	 *  specification for more details on IDENTIFY for NVMe controllers.
	 */
	cdata = spdk_nvme_ctrlr_get_data(ctrlr);

	snprintf(entry->name, sizeof(entry->name), "%-20.20s (%-20.20s)", cdata->mn, cdata->sn);

	entry->ctrlr = ctrlr;
	TAILQ_INSERT_TAIL(&g_controllers, entry, link);

	/*
	 * Each controller has one or more namespaces.  An NVMe namespace is basically
	 *  equivalent to a SCSI LUN.  The controller's IDENTIFY data tells us how
	 *  many namespaces exist on the controller.  For Intel(R) P3X00 controllers,
	 *  it will just be one namespace.
	 *
	 * Note that in NVMe, namespace IDs start at 1, not 0.
	 */
	for (nsid = spdk_nvme_ctrlr_get_first_active_ns(ctrlr); nsid != 0;
	     nsid = spdk_nvme_ctrlr_get_next_active_ns(ctrlr, nsid)) {
		ns = spdk_nvme_ctrlr_get_ns(ctrlr, nsid);
		if (ns == NULL) {
			continue;
		}
		register_ns(ctrlr, ns);
	}
}

struct ns_entry *  initialize_nvmessd(int num_server_thread , int num_poller_thread , int num_gc_thread){
    int rc , i , k , core = 0;
	struct trid_entry* trid_entry;
	struct ns_entry * ns_entry;
	struct spdk_env_opts opts ;

	char * core_mask;
	core_mask = (char * )malloc(sizeof(char)* 10);
	memset(core_mask,0,10);
	for( i = 0 , k = 1 ; i <= (num_server_thread + num_poller_thread + num_gc_thread) ; i ++){
		core += k;
		k *= 2;
	}
	snprintf(core_mask , 10 , "%x" , core);
	printf("core_mask is %s\n" , core_mask);

    printf("Initializing NVMe Controllers\n");
	spdk_env_opts_init(&opts);//初始化opts
	opts.core_mask = core_mask;
    opts.name = "initialize";
	if (spdk_env_init(&opts) < 0) {
		fprintf(stderr, "Unable to initialize SPDK env\n");
		return 0;
	}

    char * trid_initialize = "trtype:PCIe traddr:0000:83:00.0";
    add_trid(trid_initialize);
    
    trid_entry = g_trid_list.tqh_first;

    //printf("&trid_entry->trid is %d,g_trid_list is %d,g_trid_list.tqh_first is %d\n",&trid_entry->trid,&g_trid_list,g_trid_list.tqh_first);
    // printf("trid_entry->trid.traddr is %s\n",trid_entry->trid.traddr);

	if (rc = spdk_nvme_probe(&trid_entry->trid, trid_entry, probe_cb, attach_cb, NULL) != 0) {	
		printf("error while spdk_nvme_probe\n");
	}
    if (rc != 0) {
		printf("spdk_nvme_probe() failed\n");
		rc = 1;
        return 0;
	}


	if (TAILQ_EMPTY(&g_controllers)) {
		printf("no NVMe controllers found\n");
		rc = 1;
        return 0;
	}
    if (TAILQ_EMPTY(&g_namespaces)) {
		printf("no NVMe namespace found\n");
		rc = 1;
        return 0;
	}

    hello_world(num_poller_thread);


    return ns_entry;
}

**Comment by WenhaoYuan604** [comment]
Here is my code.

The main modifications were made to the read complete and write complete functions.

```
#include "./spdk_struct.h"

int add_trid(const char *trid_str)
{
struct trid_entry *trid_entry;
struct spdk_nvme_transport_id *trid;
char *ns;
char *hostnqn;

trid_entry = calloc(1, sizeof(*trid_entry));
if (trid_entry == NULL) {
    printf("trid_entry allocted memory failed\n");
	return -1;
}
trid = &trid_entry->trid;
trid->trtype = SPDK_NVME_TRANSPORT_PCIE;
snprintf(trid->subnqn, sizeof(trid->subnqn), "%s", SPDK_NVMF_DISCOVERY_NQN);
if (spdk_nvme_transport_id_parse(trid, trid_str) != 0) {
	printf("Invalid transport ID format '%s'\n", trid_str);
	free(trid_entry);
	return 1;
}
ns = strcasestr(trid_str, "ns:");
if (ns) {
	char nsid_str[6]; /* 5 digits maximum in an nsid */
	int len;
	int nsid;
	ns += 3;
	len = strcspn(ns, " \t\n");
	if (len > 5) {
		printf("NVMe namespace IDs must be 5 digits or less\n");
		free(trid_entry);
		return 1;
	}
	memcpy(nsid_str, ns, len);
	nsid_str[len] = '\0';
	nsid = spdk_strtol(nsid_str, 10);
	if (nsid <= 0 || nsid > 65535) {
		fprintf(stderr, "NVMe namespace IDs must be less than 65536 and greater than 0\n");
		free(trid_entry);
		return 1;
	}
	trid_entry->nsid = (uint16_t)nsid;
}
hostnqn = strcasestr(trid_str, "hostnqn:");
if (hostnqn) {
	size_t len;
	hostnqn += strlen("hostnqn:");
	len = strcspn(hostnqn, " \t\n");
	if (len > (sizeof(trid_entry->hostnqn) - 1)) {
		fprintf(stderr, "Host NQN is too long\n");
		free(trid_entry);
		return 1;
	}
	memcpy(trid_entry->hostnqn, hostnqn, len);
	trid_entry->hostnqn[len] = '\0';
}
// printf("insert g_trid_list %d ,trid_entry is %d\n",&g_trid_list,trid_entry);
TAILQ_INSERT_TAIL(&g_trid_list, trid_entry, tailq);
// printf("after insert g_trid_list %d ,g_trid_list.first is %d\n",&g_trid_list.tqh_first);
return 0;
}

bool
probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
struct spdk_nvme_ctrlr_opts *opts)
{
printf("Attaching to %s\n", trid->traddr);

return true;
}

void
register_ns(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_ns *ns)
{
struct ns_entry *entry;

if (!spdk_nvme_ns_is_active(ns)) {
	return;
}

entry = malloc(sizeof(struct ns_entry));
if (entry == NULL) {
	perror("ns_entry malloc");
	exit(1);
}

entry->ctrlr = ctrlr;
entry->ns = ns;
TAILQ_INSERT_TAIL(&g_namespaces, entry, link);

printf("  Namespace ID: %d size: %juGB\n", spdk_nvme_ns_get_id(ns),
       spdk_nvme_ns_get_size(ns) / 1000000000);
}

struct hello_world_sequence {
struct ns_entry *ns_entry;
char *buf;
unsigned using_cmb_io;
int is_completed;
};

static void
read_complete(void *arg, const struct spdk_nvme_cpl *completion)
{
struct hello_world_sequence *sequence = arg;

/* Assume the I/O was successful */
sequence->is_completed = 1;
/* See if an error occurred. If so, display information
 * about it, and set completion value so that I/O
 * caller is aware that an error occurred.
 */
if (spdk_nvme_cpl_is_error(completion)) {
	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
	fprintf(stderr, "Read I/O failed, aborting run\n");
	sequence->is_completed = 2;
	exit(1);
}

/*
 * The read I/O has completed.  Print the contents of the
 *  buffer, free the buffer, then mark the sequence as
 *  completed.  This will trigger the hello_world() function
 *  to exit its polling loop.
 */
char * buffer = sequence->buf + 1024;
printf("buffer is %s%s\n" , sequence->buf , buffer);

int length = 0;
for(int i = 0 ; i< 0x1000 ; i++){
	if(sequence->buf[i] != 0)
		length++;
}
printf("length is %d\n" , length);
spdk_free(sequence->buf);
}

static void
write_complete(void *arg, const struct spdk_nvme_cpl *completion)
{
struct hello_world_sequence *sequence = arg;
struct ns_entry *ns_entry = sequence->ns_entry;
int rc;

/* See if an error occurred. If so, display information
 * about it, and set completion value so that I/O
 * caller is aware that an error occurred.
 */
if (spdk_nvme_cpl_is_error(completion)) {
	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
	fprintf(stderr, "Write I/O failed, aborting run\n");
	sequence->is_completed = 2;
	exit(1);
}
/*
 * The write I/O has completed.  Free the buffer associated with
 *  the write I/O and allocate a new zeroed buffer for reading
 *  the data back from the NVMe namespace.
 */
if (sequence->using_cmb_io) {
	spdk_nvme_ctrlr_unmap_cmb(ns_entry->ctrlr);
} else {
	spdk_free(sequence->buf);
}
sequence->buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);

rc = spdk_nvme_ns_cmd_read(ns_entry->ns, ns_entry->qpair[0], sequence->buf,
			   0, /* LBA start */
			   1, /* number of LBAs */
			   read_complete, (void *)sequence, 0);
if (rc != 0) {
	fprintf(stderr, "starting read I/O failed\n");
	exit(1);
}
}

static void
hello_world(int num_qpair)
{
struct ns_entry *ns_entry;
struct hello_world_sequence sequence;
int rc;
size_t sz;

TAILQ_FOREACH(ns_entry, &g_namespaces, link) {
	
	ns_entry->qpair = (struct spdk_nvme_qpair **)malloc(sizeof(struct spdk_nvme_qpair *) * num_qpair);
	for(int i = 0 ; i < num_qpair ; i++ ){
		ns_entry->qpair[i] = spdk_nvme_ctrlr_alloc_io_qpair(ns_entry->ctrlr, NULL, 0);
		if (ns_entry->qpair[i] == NULL) {
			printf("ERROR: spdk_nvme_ctrlr_alloc_io_qpair() failed\n");
			return ;
		}
	}

	sequence.using_cmb_io = 1;
	sequence.buf = spdk_nvme_ctrlr_map_cmb(ns_entry->ctrlr, &sz);
	if (sequence.buf == NULL || sz < 0x1000) {
		sequence.using_cmb_io = 0;
		sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
	}
	if (sequence.buf == NULL) {
		printf("ERROR: write buffer allocation failed\n");
		return;
	}
	if (sequence.using_cmb_io) {
		printf("INFO: using controller memory buffer for IO\n");
	} else {
		printf("INFO: using host memory buffer for IO\n");
	}
	sequence.is_completed = 0;
	sequence.ns_entry = ns_entry;

	/*
	 * If the namespace is a Zoned Namespace, rather than a regular
	 * NVM namespace, we need to reset the first zone, before we
	 * write to it. This not needed for regular NVM namespaces.
	 */
	/*
	 * Print "Hello world!" to sequence.buf.  We will write this data to LBA
	 *  0 on the namespace, and then later read it back into a separate buffer
	 *  to demonstrate the full I/O path.
	 */
	snprintf(sequence.buf, 0x1000, "%s", "Hello world!\n");
	char * buffer = sequence.buf + 1024;
	snprintf(buffer, 0x0c00, "%s", "Hello world!\n");
	printf("buffer is %s%s\n" , sequence.buf , buffer);
	int length = 0;
	for(int i = 0 ; i< 0x1000 ; i++){
		if(sequence.buf[i] != 0)
			length++;
	}
	printf("length is %d\n" , length);
	/*
	 * Write the data buffer to LBA 0 of this namespace.  "write_complete" and
	 *  "&sequence" are specified as the completion callback function and
	 *  argument respectively.  write_complete() will be called with the
	 *  value of &sequence as a parameter when the write I/O is completed.
	 *  This allows users to potentially specify different completion
	 *  callback routines for each I/O, as well as pass a unique handle
	 *  as an argument so the application knows which I/O has completed.
	 *
	 * Note that the SPDK NVMe driver will only check for completions
	 *  when the application calls spdk_nvme_qpair_process_completions().
	 *  It is the responsibility of the application to trigger the polling
	 *  process.
	 */
	rc = spdk_nvme_ns_cmd_write(ns_entry->ns, ns_entry->qpair[0], sequence.buf,
				    0, /* LBA start */
				    1, /* number of LBAs */
				    write_complete, &sequence, SPDK_NVME_IO_FLAGS_FORCE_UNIT_ACCESS);
	if (rc != 0) {
		fprintf(stderr, "starting write I/O failed\n");
		exit(1);
	}

	/*
	 * Poll for completions.  0 here means process all available completions.
	 *  In certain usage models, the caller may specify a positive integer
	 *  instead of 0 to signify the maximum number of completions it should
	 *  process.  This function will never block - if there are no
	 *  completions pending on the specified qpair, it will return immediately.
	 *
	 * When the write I/O completes, write_complete() will submit a new I/O
	 *  to read LBA 0 into a separate buffer, specifying read_complete() as its
	 *  completion routine.  When the read I/O completes, read_complete() will
	 *  print the buffer contents and set sequence.is_completed = 1.  That will
	 *  break this loop and then exit the program.
	 */
	while (!sequence.is_completed) {
		spdk_nvme_qpair_process_completions(ns_entry->qpair[0], 0);
	}

	/*
	 * Free the I/O qpair.  This typically is done when an application exits.
	 *  But SPDK does support freeing and then reallocating qpairs during
	 *  operation.  It is the responsibility of the caller to ensure all
	 *  pending I/O are completed before trying to free the qpair.
	 */
	// spdk_nvme_ctrlr_free_io_qpair(ns_entry->qpair);
}
}

void
attach_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
struct spdk_nvme_ctrlr *ctrlr, const struct spdk_nvme_ctrlr_opts *opts)
{
int nsid;
struct ctrlr_entry *entry;
struct spdk_nvme_ns *ns;
const struct spdk_nvme_ctrlr_data *cdata;

entry = malloc(sizeof(struct ctrlr_entry));
if (entry == NULL) {
	perror("ctrlr_entry malloc");
	exit(1);
}

printf("Attached to %s\n", trid->traddr);

/*
 * spdk_nvme_ctrlr is the logical abstraction in SPDK for an NVMe
 *  controller.  During initialization, the IDENTIFY data for the
 *  controller is read using an NVMe admin command, and that data
 *  can be retrieved using spdk_nvme_ctrlr_get_data() to get
 *  detailed information on the controller.  Refer to the NVMe
 *  specification for more details on IDENTIFY for NVMe controllers.
 */
cdata = spdk_nvme_ctrlr_get_data(ctrlr);

snprintf(entry->name, sizeof(entry->name), "%-20.20s (%-20.20s)", cdata->mn, cdata->sn);

entry->ctrlr = ctrlr;
TAILQ_INSERT_TAIL(&g_controllers, entry, link);

/*
 * Each controller has one or more namespaces.  An NVMe namespace is basically
 *  equivalent to a SCSI LUN.  The controller's IDENTIFY data tells us how
 *  many namespaces exist on the controller.  For Intel(R) P3X00 controllers,
 *  it will just be one namespace.
 *
 * Note that in NVMe, namespace IDs start at 1, not 0.
 */
for (nsid = spdk_nvme_ctrlr_get_first_active_ns(ctrlr); nsid != 0;
     nsid = spdk_nvme_ctrlr_get_next_active_ns(ctrlr, nsid)) {
	ns = spdk_nvme_ctrlr_get_ns(ctrlr, nsid);
	if (ns == NULL) {
		continue;
	}
	register_ns(ctrlr, ns);
}
}

struct ns_entry * initialize_nvmessd(int num_server_thread , int num_poller_thread , int num_gc_thread){
int rc , i , k , core = 0;
struct trid_entry* trid_entry;
struct ns_entry * ns_entry;
struct spdk_env_opts opts ;

char * core_mask;
core_mask = (char * )malloc(sizeof(char)* 10);
memset(core_mask,0,10);
for( i = 0 , k = 1 ; i <= (num_server_thread + num_poller_thread + num_gc_thread) ; i ++){
	core += k;
	k *= 2;
}
snprintf(core_mask , 10 , "%x" , core);
printf("core_mask is %s\n" , core_mask);

printf("Initializing NVMe Controllers\n");
spdk_env_opts_init(&opts);//初始化opts
opts.core_mask = core_mask;
opts.name = "initialize";
if (spdk_env_init(&opts) < 0) {
	fprintf(stderr, "Unable to initialize SPDK env\n");
	return 0;
}

char * trid_initialize = "trtype:PCIe traddr:0000:83:00.0";
add_trid(trid_initialize);

trid_entry = g_trid_list.tqh_first;

//printf("&trid_entry->trid is %d,g_trid_list is %d,g_trid_list.tqh_first is %d\n",&trid_entry->trid,&g_trid_list,g_trid_list.tqh_first);
// printf("trid_entry->trid.traddr is %s\n",trid_entry->trid.traddr);

if (rc = spdk_nvme_probe(&trid_entry->trid, trid_entry, probe_cb, attach_cb, NULL) != 0) {	
	printf("error while spdk_nvme_probe\n");
}
if (rc != 0) {
	printf("spdk_nvme_probe() failed\n");
	rc = 1;
    return 0;
}


if (TAILQ_EMPTY(&g_controllers)) {
	printf("no NVMe controllers found\n");
	rc = 1;
    return 0;
}
if (TAILQ_EMPTY(&g_namespaces)) {
	printf("no NVMe namespace found\n");
	rc = 1;
    return 0;
}

hello_world(num_poller_thread);


return ns_entry;
```

**Comment by WenhaoYuan604** [comment]
> Here is my code.
> 
> The main modifications were made to the read complete and write complete functions.
> 
> ```
> #include "./spdk_struct.h"
> 
> int add_trid(const char *trid_str)
> {
> struct trid_entry *trid_entry;
> struct spdk_nvme_transport_id *trid;
> char *ns;
> char *hostnqn;
> 
> trid_entry = calloc(1, sizeof(*trid_entry));
> if (trid_entry == NULL) {
>     printf("trid_entry allocted memory failed\n");
> 	return -1;
> }
> trid = &trid_entry->trid;
> trid->trtype = SPDK_NVME_TRANSPORT_PCIE;
> snprintf(trid->subnqn, sizeof(trid->subnqn), "%s", SPDK_NVMF_DISCOVERY_NQN);
> if (spdk_nvme_transport_id_parse(trid, trid_str) != 0) {
> 	printf("Invalid transport ID format '%s'\n", trid_str);
> 	free(trid_entry);
> 	return 1;
> }
> ns = strcasestr(trid_str, "ns:");
> if (ns) {
> 	char nsid_str[6]; /* 5 digits maximum in an nsid */
> 	int len;
> 	int nsid;
> 	ns += 3;
> 	len = strcspn(ns, " \t\n");
> 	if (len > 5) {
> 		printf("NVMe namespace IDs must be 5 digits or less\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	memcpy(nsid_str, ns, len);
> 	nsid_str[len] = '\0';
> 	nsid = spdk_strtol(nsid_str, 10);
> 	if (nsid <= 0 || nsid > 65535) {
> 		fprintf(stderr, "NVMe namespace IDs must be less than 65536 and greater than 0\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	trid_entry->nsid = (uint16_t)nsid;
> }
> hostnqn = strcasestr(trid_str, "hostnqn:");
> if (hostnqn) {
> 	size_t len;
> 	hostnqn += strlen("hostnqn:");
> 	len = strcspn(hostnqn, " \t\n");
> 	if (len > (sizeof(trid_entry->hostnqn) - 1)) {
> 		fprintf(stderr, "Host NQN is too long\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	memcpy(trid_entry->hostnqn, hostnqn, len);
> 	trid_entry->hostnqn[len] = '\0';
> }
> // printf("insert g_trid_list %d ,trid_entry is %d\n",&g_trid_list,trid_entry);
> TAILQ_INSERT_TAIL(&g_trid_list, trid_entry, tailq);
> // printf("after insert g_trid_list %d ,g_trid_list.first is %d\n",&g_trid_list.tqh_first);
> return 0;
> }
> 
> bool
> probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
> struct spdk_nvme_ctrlr_opts *opts)
> {
> printf("Attaching to %s\n", trid->traddr);
> 
> return true;
> }
> 
> void
> register_ns(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_ns *ns)
> {
> struct ns_entry *entry;
> 
> if (!spdk_nvme_ns_is_active(ns)) {
> 	return;
> }
> 
> entry = malloc(sizeof(struct ns_entry));
> if (entry == NULL) {
> 	perror("ns_entry malloc");
> 	exit(1);
> }
> 
> entry->ctrlr = ctrlr;
> entry->ns = ns;
> TAILQ_INSERT_TAIL(&g_namespaces, entry, link);
> 
> printf("  Namespace ID: %d size: %juGB\n", spdk_nvme_ns_get_id(ns),
>        spdk_nvme_ns_get_size(ns) / 1000000000);
> }
> 
> struct hello_world_sequence {
> struct ns_entry *ns_entry;
> char *buf;
> unsigned using_cmb_io;
> int is_completed;
> };
> 
> static void
> read_complete(void *arg, const struct spdk_nvme_cpl *completion)
> {
> struct hello_world_sequence *sequence = arg;
> 
> /* Assume the I/O was successful */
> sequence->is_completed = 1;
> /* See if an error occurred. If so, display information
>  * about it, and set completion value so that I/O
>  * caller is aware that an error occurred.
>  */
> if (spdk_nvme_cpl_is_error(completion)) {
> 	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
> 	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
> 	fprintf(stderr, "Read I/O failed, aborting run\n");
> 	sequence->is_completed = 2;
> 	exit(1);
> }
> 
> /*
>  * The read I/O has completed.  Print the contents of the
>  *  buffer, free the buffer, then mark the sequence as
>  *  completed.  This will trigger the hello_world() function
>  *  to exit its polling loop.
>  */
> char * buffer = sequence->buf + 1024;
> printf("buffer is %s%s\n" , sequence->buf , buffer);
> 
> int length = 0;
> for(int i = 0 ; i< 0x1000 ; i++){
> 	if(sequence->buf[i] != 0)
> 		length++;
> }
> printf("length is %d\n" , length);
> spdk_free(sequence->buf);
> }
> 
> static void
> write_complete(void *arg, const struct spdk_nvme_cpl *completion)
> {
> struct hello_world_sequence *sequence = arg;
> struct ns_entry *ns_entry = sequence->ns_entry;
> int rc;
> 
> /* See if an error occurred. If so, display information
>  * about it, and set completion value so that I/O
>  * caller is aware that an error occurred.
>  */
> if (spdk_nvme_cpl_is_error(completion)) {
> 	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
> 	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
> 	fprintf(stderr, "Write I/O failed, aborting run\n");
> 	sequence->is_completed = 2;
> 	exit(1);
> }
> /*
>  * The write I/O has completed.  Free the buffer associated with
>  *  the write I/O and allocate a new zeroed buffer for reading
>  *  the data back from the NVMe namespace.
>  */
> if (sequence->using_cmb_io) {
> 	spdk_nvme_ctrlr_unmap_cmb(ns_entry->ctrlr);
> } else {
> 	spdk_free(sequence->buf);
> }
> sequence->buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
> 
> rc = spdk_nvme_ns_cmd_read(ns_entry->ns, ns_entry->qpair[0], sequence->buf,
> 			   0, /* LBA start */
> 			   1, /* number of LBAs */
> 			   read_complete, (void *)sequence, 0);
> if (rc != 0) {
> 	fprintf(stderr, "starting read I/O failed\n");
> 	exit(1);
> }
> }
> 
> static void
> hello_world(int num_qpair)
> {
> struct ns_entry *ns_entry;
> struct hello_world_sequence sequence;
> int rc;
> size_t sz;
> 
> TAILQ_FOREACH(ns_entry, &g_namespaces, link) {
> 	
> 	ns_entry->qpair = (struct spdk_nvme_qpair **)malloc(sizeof(struct spdk_nvme_qpair *) * num_qpair);
> 	for(int i = 0 ; i < num_qpair ; i++ ){
> 		ns_entry->qpair[i] = spdk_nvme_ctrlr_alloc_io_qpair(ns_entry->ctrlr, NULL, 0);
> 		if (ns_entry->qpair[i] == NULL) {
> 			printf("ERROR: spdk_nvme_ctrlr_alloc_io_qpair() failed\n");
> 			return ;
> 		}
> 	}
> 
> 	sequence.using_cmb_io = 1;
> 	sequence.buf = spdk_nvme_ctrlr_map_cmb(ns_entry->ctrlr, &sz);
> 	if (sequence.buf == NULL || sz < 0x1000) {
> 		sequence.using_cmb_io = 0;
> 		sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
> 	}
> 	if (sequence.buf == NULL) {
> 		printf("ERROR: write buffer allocation failed\n");
> 		return;
> 	}
> 	if (sequence.using_cmb_io) {
> 		printf("INFO: using controller memory buffer for IO\n");
> 	} else {
> 		printf("INFO: using host memory buffer for IO\n");
> 	}
> 	sequence.is_completed = 0;
> 	sequence.ns_entry = ns_entry;
> 
> 	/*
> 	 * If the namespace is a Zoned Namespace, rather than a regular
> 	 * NVM namespace, we need to reset the first zone, before we
> 	 * write to it. This not needed for regular NVM namespaces.
> 	 */
> 	/*
> 	 * Print "Hello world!" to sequence.buf.  We will write this data to LBA
> 	 *  0 on the namespace, and then later read it back into a separate buffer
> 	 *  to demonstrate the full I/O path.
> 	 */
> 	snprintf(sequence.buf, 0x1000, "%s", "Hello world!\n");
> 	char * buffer = sequence.buf + 1024;
> 	snprintf(buffer, 0x0c00, "%s", "Hello world!\n");
> 	printf("buffer is %s%s\n" , sequence.buf , buffer);
> 	int length = 0;
> 	for(int i = 0 ; i< 0x1000 ; i++){
> 		if(sequence.buf[i] != 0)
> 			length++;
> 	}
> 	printf("length is %d\n" , length);
> 	/*
> 	 * Write the data buffer to LBA 0 of this namespace.  "write_complete" and
> 	 *  "&sequence" are specified as the completion callback function and
> 	 *  argument respectively.  write_complete() will be called with the
> 	 *  value of &sequence as a parameter when the write I/O is completed.
> 	 *  This allows users to potentially specify different completion
> 	 *  callback routines for each I/O, as well as pass a unique handle
> 	 *  as an argument so the application knows which I/O has completed.
> 	 *
> 	 * Note that the SPDK NVMe driver will only check for completions
> 	 *  when the application calls spdk_nvme_qpair_process_completions().
> 	 *  It is the responsibility of the application to trigger the polling
> 	 *  process.
> 	 */
> 	rc = spdk_nvme_ns_cmd_write(ns_entry->ns, ns_entry->qpair[0], sequence.buf,
> 				    0, /* LBA start */
> 				    1, /* number of LBAs */
> 				    write_complete, &sequence, SPDK_NVME_IO_FLAGS_FORCE_UNIT_ACCESS);
> 	if (rc != 0) {
> 		fprintf(stderr, "starting write I/O failed\n");
> 		exit(1);
> 	}
> 
> 	/*
> 	 * Poll for completions.  0 here means process all available completions.
> 	 *  In certain usage models, the caller may specify a positive integer
> 	 *  instead of 0 to signify the maximum number of completions it should
> 	 *  process.  This function will never block - if there are no
> 	 *  completions pending on the specified qpair, it will return immediately.
> 	 *
> 	 * When the write I/O completes, write_complete() will submit a new I/O
> 	 *  to read LBA 0 into a separate buffer, specifying read_complete() as its
> 	 *  completion routine.  When the read I/O completes, read_complete() will
> 	 *  print the buffer contents and set sequence.is_completed = 1.  That will
> 	 *  break this loop and then exit the program.
> 	 */
> 	while (!sequence.is_completed) {
> 		spdk_nvme_qpair_process_completions(ns_entry->qpair[0], 0);
> 	}
> 
> 	/*
> 	 * Free the I/O qpair.  This typically is done when an application exits.
> 	 *  But SPDK does support freeing and then reallocating qpairs during
> 	 *  operation.  It is the responsibility of the caller to ensure all
> 	 *  pending I/O are completed before trying to free the qpair.
> 	 */
> 	// spdk_nvme_ctrlr_free_io_qpair(ns_entry->qpair);
> }
> }
> 
> void
> attach_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
> struct spdk_nvme_ctrlr *ctrlr, const struct spdk_nvme_ctrlr_opts *opts)
> {
> int nsid;
> struct ctrlr_entry *entry;
> struct spdk_nvme_ns *ns;
> const struct spdk_nvme_ctrlr_data *cdata;
> 
> entry = malloc(sizeof(struct ctrlr_entry));
> if (entry == NULL) {
> 	perror("ctrlr_entry malloc");
> 	exit(1);
> }
> 
> printf("Attached to %s\n", trid->traddr);
> 
> /*
>  * spdk_nvme_ctrlr is the logical abstraction in SPDK for an NVMe
>  *  controller.  During initialization, the IDENTIFY data for the
>  *  controller is read using an NVMe admin command, and that data
>  *  can be retrieved using spdk_nvme_ctrlr_get_data() to get
>  *  detailed information on the controller.  Refer to the NVMe
>  *  specification for more details on IDENTIFY for NVMe controllers.
>  */
> cdata = spdk_nvme_ctrlr_get_data(ctrlr);
> 
> snprintf(entry->name, sizeof(entry->name), "%-20.20s (%-20.20s)", cdata->mn, cdata->sn);
> 
> entry->ctrlr = ctrlr;
> TAILQ_INSERT_TAIL(&g_controllers, entry, link);
> 
> /*
>  * Each controller has one or more namespaces.  An NVMe namespace is basically
>  *  equivalent to a SCSI LUN.  The controller's IDENTIFY data tells us how
>  *  many namespaces exist on the controller.  For Intel(R) P3X00 controllers,
>  *  it will just be one namespace.
>  *
>  * Note that in NVMe, namespace IDs start at 1, not 0.
>  */
> for (nsid = spdk_nvme_ctrlr_get_first_active_ns(ctrlr); nsid != 0;
>      nsid = spdk_nvme_ctrlr_get_next_active_ns(ctrlr, nsid)) {
> 	ns = spdk_nvme_ctrlr_get_ns(ctrlr, nsid);
> 	if (ns == NULL) {
> 		continue;
> 	}
> 	register_ns(ctrlr, ns);
> }
> }
> 
> struct ns_entry * initialize_nvmessd(int num_server_thread , int num_poller_thread , int num_gc_thread){
> int rc , i , k , core = 0;
> struct trid_entry* trid_entry;
> struct ns_entry * ns_entry;
> struct spdk_env_opts opts ;
> 
> char * core_mask;
> core_mask = (char * )malloc(sizeof(char)* 10);
> memset(core_mask,0,10);
> for( i = 0 , k = 1 ; i <= (num_server_thread + num_poller_thread + num_gc_thread) ; i ++){
> 	core += k;
> 	k *= 2;
> }
> snprintf(core_mask , 10 , "%x" , core);
> printf("core_mask is %s\n" , core_mask);
> 
> printf("Initializing NVMe Controllers\n");
> spdk_env_opts_init(&opts);//初始化opts
> opts.core_mask = core_mask;
> opts.name = "initialize";
> if (spdk_env_init(&opts) < 0) {
> 	fprintf(stderr, "Unable to initialize SPDK env\n");
> 	return 0;
> }
> 
> char * trid_initialize = "trtype:PCIe traddr:0000:83:00.0";
> add_trid(trid_initialize);
> 
> trid_entry = g_trid_list.tqh_first;
> 
> //printf("&trid_entry->trid is %d,g_trid_list is %d,g_trid_list.tqh_first is %d\n",&trid_entry->trid,&g_trid_list,g_trid_list.tqh_first);
> // printf("trid_entry->trid.traddr is %s\n",trid_entry->trid.traddr);
> 
> if (rc = spdk_nvme_probe(&trid_entry->trid, trid_entry, probe_cb, attach_cb, NULL) != 0) {	
> 	printf("error while spdk_nvme_probe\n");
> }
> if (rc != 0) {
> 	printf("spdk_nvme_probe() failed\n");
> 	rc = 1;
>     return 0;
> }
> 
> 
> if (TAILQ_EMPTY(&g_controllers)) {
> 	printf("no NVMe controllers found\n");
> 	rc = 1;
>     return 0;
> }
> if (TAILQ_EMPTY(&g_namespaces)) {
> 	printf("no NVMe namespace found\n");
> 	rc = 1;
>     return 0;
> }
> 
> hello_world(num_poller_thread);
> 
> 
> return ns_entry;
> ```
Here is output:

core_mask is 7
Initializing NVMe Controllers
TELEMETRY: No legacy callbacks, legacy socket not created
Attaching to 0000:83:00.0
Attached to 0000:83:00.0
  Namespace ID: 1 size: 1000GB
INFO: using host memory buffer for IO
buffer is Hello world!
Hello world!

length is 26
buffer is Hello world!

length is 13

**Comment by jimharris** [comment]
> Here is my code.
> 
> The main modifications were made to the read complete and write complete functions.
> 
> ```
> #include "./spdk_struct.h"
> 
> int add_trid(const char *trid_str)
> {
> struct trid_entry *trid_entry;
> struct spdk_nvme_transport_id *trid;
> char *ns;
> char *hostnqn;
> 
> trid_entry = calloc(1, sizeof(*trid_entry));
> if (trid_entry == NULL) {
>     printf("trid_entry allocted memory failed\n");
> 	return -1;
> }
> trid = &trid_entry->trid;
> trid->trtype = SPDK_NVME_TRANSPORT_PCIE;
> snprintf(trid->subnqn, sizeof(trid->subnqn), "%s", SPDK_NVMF_DISCOVERY_NQN);
> if (spdk_nvme_transport_id_parse(trid, trid_str) != 0) {
> 	printf("Invalid transport ID format '%s'\n", trid_str);
> 	free(trid_entry);
> 	return 1;
> }
> ns = strcasestr(trid_str, "ns:");
> if (ns) {
> 	char nsid_str[6]; /* 5 digits maximum in an nsid */
> 	int len;
> 	int nsid;
> 	ns += 3;
> 	len = strcspn(ns, " \t\n");
> 	if (len > 5) {
> 		printf("NVMe namespace IDs must be 5 digits or less\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	memcpy(nsid_str, ns, len);
> 	nsid_str[len] = '\0';
> 	nsid = spdk_strtol(nsid_str, 10);
> 	if (nsid <= 0 || nsid > 65535) {
> 		fprintf(stderr, "NVMe namespace IDs must be less than 65536 and greater than 0\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	trid_entry->nsid = (uint16_t)nsid;
> }
> hostnqn = strcasestr(trid_str, "hostnqn:");
> if (hostnqn) {
> 	size_t len;
> 	hostnqn += strlen("hostnqn:");
> 	len = strcspn(hostnqn, " \t\n");
> 	if (len > (sizeof(trid_entry->hostnqn) - 1)) {
> 		fprintf(stderr, "Host NQN is too long\n");
> 		free(trid_entry);
> 		return 1;
> 	}
> 	memcpy(trid_entry->hostnqn, hostnqn, len);
> 	trid_entry->hostnqn[len] = '\0';
> }
> // printf("insert g_trid_list %d ,trid_entry is %d\n",&g_trid_list,trid_entry);
> TAILQ_INSERT_TAIL(&g_trid_list, trid_entry, tailq);
> // printf("after insert g_trid_list %d ,g_trid_list.first is %d\n",&g_trid_list.tqh_first);
> return 0;
> }
> 
> bool
> probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
> struct spdk_nvme_ctrlr_opts *opts)
> {
> printf("Attaching to %s\n", trid->traddr);
> 
> return true;
> }
> 
> void
> register_ns(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_ns *ns)
> {
> struct ns_entry *entry;
> 
> if (!spdk_nvme_ns_is_active(ns)) {
> 	return;
> }
> 
> entry = malloc(sizeof(struct ns_entry));
> if (entry == NULL) {
> 	perror("ns_entry malloc");
> 	exit(1);
> }
> 
> entry->ctrlr = ctrlr;
> entry->ns = ns;
> TAILQ_INSERT_TAIL(&g_namespaces, entry, link);
> 
> printf("  Namespace ID: %d size: %juGB\n", spdk_nvme_ns_get_id(ns),
>        spdk_nvme_ns_get_size(ns) / 1000000000);
> }
> 
> struct hello_world_sequence {
> struct ns_entry *ns_entry;
> char *buf;
> unsigned using_cmb_io;
> int is_completed;
> };
> 
> static void
> read_complete(void *arg, const struct spdk_nvme_cpl *completion)
> {
> struct hello_world_sequence *sequence = arg;
> 
> /* Assume the I/O was successful */
> sequence->is_completed = 1;
> /* See if an error occurred. If so, display information
>  * about it, and set completion value so that I/O
>  * caller is aware that an error occurred.
>  */
> if (spdk_nvme_cpl_is_error(completion)) {
> 	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
> 	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
> 	fprintf(stderr, "Read I/O failed, aborting run\n");
> 	sequence->is_completed = 2;
> 	exit(1);
> }
> 
> /*
>  * The read I/O has completed.  Print the contents of the
>  *  buffer, free the buffer, then mark the sequence as
>  *  completed.  This will trigger the hello_world() function
>  *  to exit its polling loop.
>  */
> char * buffer = sequence->buf + 1024;
> printf("buffer is %s%s\n" , sequence->buf , buffer);
> 
> int length = 0;
> for(int i = 0 ; i< 0x1000 ; i++){
> 	if(sequence->buf[i] != 0)
> 		length++;
> }
> printf("length is %d\n" , length);
> spdk_free(sequence->buf);
> }
> 
> static void
> write_complete(void *arg, const struct spdk_nvme_cpl *completion)
> {
> struct hello_world_sequence *sequence = arg;
> struct ns_entry *ns_entry = sequence->ns_entry;
> int rc;
> 
> /* See if an error occurred. If so, display information
>  * about it, and set completion value so that I/O
>  * caller is aware that an error occurred.
>  */
> if (spdk_nvme_cpl_is_error(completion)) {
> 	spdk_nvme_qpair_print_completion(sequence->ns_entry->qpair[0], (struct spdk_nvme_cpl *)completion);
> 	fprintf(stderr, "I/O error status: %s\n", spdk_nvme_cpl_get_status_string(&completion->status));
> 	fprintf(stderr, "Write I/O failed, aborting run\n");
> 	sequence->is_completed = 2;
> 	exit(1);
> }
> /*
>  * The write I/O has completed.  Free the buffer associated with
>  *  the write I/O and allocate a new zeroed buffer for reading
>  *  the data back from the NVMe namespace.
>  */
> if (sequence->using_cmb_io) {
> 	spdk_nvme_ctrlr_unmap_cmb(ns_entry->ctrlr);
> } else {
> 	spdk_free(sequence->buf);
> }
> sequence->buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
> 
> rc = spdk_nvme_ns_cmd_read(ns_entry->ns, ns_entry->qpair[0], sequence->buf,
> 			   0, /* LBA start */
> 			   1, /* number of LBAs */
> 			   read_complete, (void *)sequence, 0);
> if (rc != 0) {
> 	fprintf(stderr, "starting read I/O failed\n");
> 	exit(1);
> }
> }
> 
> static void
> hello_world(int num_qpair)
> {
> struct ns_entry *ns_entry;
> struct hello_world_sequence sequence;
> int rc;
> size_t sz;
> 
> TAILQ_FOREACH(ns_entry, &g_namespaces, link) {
> 	
> 	ns_entry->qpair = (struct spdk_nvme_qpair **)malloc(sizeof(struct spdk_nvme_qpair *) * num_qpair);
> 	for(int i = 0 ; i < num_qpair ; i++ ){
> 		ns_entry->qpair[i] = spdk_nvme_ctrlr_alloc_io_qpair(ns_entry->ctrlr, NULL, 0);
> 		if (ns_entry->qpair[i] == NULL) {
> 			printf("ERROR: spdk_nvme_ctrlr_alloc_io_qpair() failed\n");
> 			return ;
> 		}
> 	}
> 
> 	sequence.using_cmb_io = 1;
> 	sequence.buf = spdk_nvme_ctrlr_map_cmb(ns_entry->ctrlr, &sz);
> 	if (sequence.buf == NULL || sz < 0x1000) {
> 		sequence.using_cmb_io = 0;
> 		sequence.buf = spdk_zmalloc(0x1000, 0x1000, NULL, SPDK_ENV_SOCKET_ID_ANY, SPDK_MALLOC_DMA);
> 	}
> 	if (sequence.buf == NULL) {
> 		printf("ERROR: write buffer allocation failed\n");
> 		return;
> 	}
> 	if (sequence.using_cmb_io) {
> 		printf("INFO: using controller memory buffer for IO\n");
> 	} else {
> 		printf("INFO: using host memory buffer for IO\n");
> 	}
> 	sequence.is_completed = 0;
> 	sequence.ns_entry = ns_entry;
> 
> 	/*
> 	 * If the namespace is a Zoned Namespace, rather than a regular
> 	 * NVM namespace, we need to reset the first zone, before we
> 	 * write to it. This not needed for regular NVM namespaces.
> 	 */
> 	/*
> 	 * Print "Hello world!" to sequence.buf.  We will write this data to LBA
> 	 *  0 on the namespace, and then later read it back into a separate buffer
> 	 *  to demonstrate the full I/O path.
> 	 */
> 	snprintf(sequence.buf, 0x1000, "%s", "Hello world!\n");
> 	char * buffer = sequence.buf + 1024;
> 	snprintf(buffer, 0x0c00, "%s", "Hello world!\n");

Here you are writing the second string into offset 1024 of the buffer.

> 	printf("buffer is %s%s\n" , sequence.buf , buffer);
> 	int length = 0;
> 	for(int i = 0 ; i< 0x1000 ; i++){
> 		if(sequence.buf[i] != 0)
> 			length++;
> 	}
> 	printf("length is %d\n" , length);
> 	/*
> 	 * Write the data buffer to LBA 0 of this namespace.  "write_complete" and
> 	 *  "&sequence" are specified as the completion callback function and
> 	 *  argument respectively.  write_complete() will be called with the
> 	 *  value of &sequence as a parameter when the write I/O is completed.
> 	 *  This allows users to potentially specify different completion
> 	 *  callback routines for each I/O, as well as pass a unique handle
> 	 *  as an argument so the application knows which I/O has completed.
> 	 *
> 	 * Note that the SPDK NVMe driver will only check for completions
> 	 *  when the application calls spdk_nvme_qpair_process_completions().
> 	 *  It is the responsibility of the application to trigger the polling
> 	 *  process.
> 	 */
> 	rc = spdk_nvme_ns_cmd_write(ns_entry->ns, ns_entry->qpair[0], sequence.buf,
> 				    0, /* LBA start */
> 				    1, /* number of LBAs */
> 				    write_complete, &sequence, SPDK_NVME_IO_FLAGS_FORCE_UNIT_ACCESS);

But here you only write 1 LBA. So it only writes the first 512 bytes.
